staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered_test.go

```diff
@@ -26,83 +25,95 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
-	"k8s.io/apimachinery/pkg/runtime"
 	auditinternal "k8s.io/apiserver/pkg/apis/audit"
-	auditv1alpha1 "k8s.io/apiserver/pkg/apis/audit/v1alpha1"
+	"k8s.io/apiserver/pkg/audit"
 )
 
-func TestBatchWebhookMaxEventsV1Alpha1(t *testing.T) {
+func newBufferedBackend() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+func newBufferedBackendWithWaitGroup() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	fakeBackend.wg = new(sync.WaitGroup)
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+func newBufferedBackendWithChannel() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	fakeBackend.ch = make(chan struct{})
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+// waitForEmptyBuffer indicates when the sendBatchEvents method has read from the
+// existing buffer. This lets test coordinate closing a timer and stop channel
+// until the for loop has read from the buffer.
+func waitForEmptyBuffer(b *bufferedBackend) {
+	for len(b.buffer) != 0 {
+		time.Sleep(time.Millisecond)
+	}
+}
+
+func TestBatchMaxEvents(t *testing.T) {
 	nRest := 10
 	events := make([]*auditinternal.Event, defaultBatchMaxSize+nRest) // greater than max size.
 	for i := range events {
 		events[i] = &auditinternal.Event{}
 	}
 
-	got := make(chan int, 2)
-	s := httptest.NewServer(newWebhookHandler(t, &auditv1alpha1.EventList{}, func(events runtime.Object) {
-		got <- len(events.(*auditv1alpha1.EventList).Items)
-	}))
-	defer s.Close()
+	stopCh := make(chan struct{})
+	timer := make(chan time.Time, 1)
 
-	backend := newTestBatchWebhook(t, s.URL, auditv1alpha1.SchemeGroupVersion)
+	backend, realBackend := newBufferedBackend()
+	backend.stopCh = stopCh
 
 	backend.ProcessEvents(events...)
 
-	stopCh := make(chan struct{})
-	timer := make(chan time.Time, 1)
-
-	backend.sendBatchEvents(backend.collectEvents(stopCh, timer))
-	require.Equal(t, defaultBatchMaxSize, <-got, "did not get batch max size")
+	backend.backend.ProcessEvents(backend.collectEvents(timer)...)
```

Comment:
You shouldn't be calling this directly here. Instead, start the backend (`backend.Run(...)`), and wait for the expected result, i.e. `realBackend.wg.Wait()`

(except I think wg.wait will give you a race condition - I recommend using a channel instead. For the fake backend, process events just puts the event slice into the channel, which is type `chan []*auditinternal.Event`).

This is preferred because it tests the backend interface, not the internal details.

Reply:
I donot think so, UT is for internal detail test. And prefer code/branch coverage. Not only interface test.

Reply:
And also interface test is covered in `TestBatchRun`.  So I do not think it deserves much more effort.